### ISSUE TITLE
tint2 and xbattmon fail to show battery status (laptop)

### ISSUE THREAD

**Issue by nkcyham** [voidlinux/void-packages#4142]
Hi,

After a fresh install of void on my laptop, I can't get tint2 to display battery status. Tried `xbattmon` and it also fails. I can only get some info out of `acpi`:

```
[0746][neoklis@okapi:~]> acpi
Battery 0: Discharging, 57%, rate information unavailable
```

xbattmon seems unable to open the relevant files in sys/:
```
[0746][neoklis@okapi:~]> xbattmon 
xbattmon: fopen /sys/class/power_supply/BAT0/capacity: No such file or directory
xbattmon: fopen /sys/class/power_supply/BAT0/capacity: No such file or directory
xbattmon: fopen /sys/class/power_supply/BAT0/capacity: No such file or directory
xbattmon: fopen /sys/class/power_supply/AC/online: No such file or directory
```

But ls shows the files to be there, and I can open them easily with gvim:
```
[1423][root@okapi:/home/neoklis]> ls -l /sys/class/power_supply/BAT0/
total 0
-rw-r--r-- 1 root root 4096 May  6 14:24 alarm
-r--r--r-- 1 root root 4096 May  6 14:24 capacity
-r--r--r-- 1 root root 4096 May  6 14:24 capacity_level
-r--r--r-- 1 root root 4096 May  6 14:24 charge_full
-r--r--r-- 1 root root 4096 May  6 14:24 charge_full_design
-r--r--r-- 1 root root 4096 May  6 14:24 charge_now
-r--r--r-- 1 root root 4096 May  6 14:24 current_now
-r--r--r-- 1 root root 4096 May  6 14:24 cycle_count
lrwxrwxrwx 1 root root    0 May  6 14:24 device -> ../../../PNP0C0A:00
-r--r--r-- 1 root root 4096 May  6 14:24 manufacturer
-r--r--r-- 1 root root 4096 May  6 14:24 model_name
drwxr-xr-x 2 root root    0 May  6 14:24 power
-r--r--r-- 1 root root 4096 May  6 14:24 present
-r--r--r-- 1 root root 4096 May  6 14:24 serial_number
-r--r--r-- 1 root root 4096 May  6 14:24 status
lrwxrwxrwx 1 root root    0 May  6 14:24 subsystem -> ../../../../../../class/power_supply
-r--r--r-- 1 root root 4096 May  6 14:24 technology
-r--r--r-- 1 root root 4096 May  6 14:24 type
-rw-r--r-- 1 root root 4096 May  6 14:23 uevent
-r--r--r-- 1 root root 4096 May  6 14:24 voltage_min_design
-r--r--r-- 1 root root 4096 May  6 14:24 voltage_now
```

My thanks in advance for any advice.

**Comment by Vaelatern** [comment]
Edited your question to use backticks, by the way. Useful for setting aside data.

As a regular user I am able to list those directories. Are you sure you have your directory level permissions right?
```
└» ls -l /sys/class/power_supply/BAT0/
total 0
-rw-r--r-- 1 root root 4096 May  6 11:18 alarm
-r--r--r-- 1 root root 4096 May  4 18:33 capacity
-r--r--r-- 1 root root 4096 May  6 11:18 capacity_level
-r--r--r-- 1 root root 4096 May  6 11:18 cycle_count
lrwxrwxrwx 1 root root    0 May  6 11:18 device -> ../../../PNP0C0A:00
-r--r--r-- 1 root root 4096 May  4 18:33 energy_full
-r--r--r-- 1 root root 4096 May  4 18:33 energy_full_design
-r--r--r-- 1 root root 4096 May  4 18:33 energy_now
-r--r--r-- 1 root root 4096 May  4 18:33 manufacturer
-r--r--r-- 1 root root 4096 May  4 18:33 model_name
drwxr-xr-x 2 root root    0 May  6 11:18 power
-r--r--r-- 1 root root 4096 May  4 18:33 power_now
-r--r--r-- 1 root root 4096 May  4 18:33 present
-r--r--r-- 1 root root 4096 May  4 18:33 serial_number
-r--r--r-- 1 root root 4096 May  4 18:33 status
lrwxrwxrwx 1 root root    0 May  4 12:55 subsystem -> ../../../../../../../../../class/power_supply
-r--r--r-- 1 root root 4096 May  4 18:33 technology
-r--r--r-- 1 root root 4096 May  4 18:33 type
-rw-r--r-- 1 root root 4096 May  4 12:55 uevent
-r--r--r-- 1 root root 4096 May  4 18:33 voltage_min_design
-r--r--r-- 1 root root 4096 May  4 18:33 voltage_now
```

Permissions on the way up are as follows
```
dr-xr-xr-x  13 root root    0 May  4 12:55 sys
drwxr-xr-x  63 root root 0 May  4 12:55 class
drwxr-xr-x 2 root root 0 May  4 12:55 power_supply

lrwxrwxrwx 1 root root 0 May  4 12:55 AC -> ../../devices/LNXSYSTM:00/LNXSYBUS:00/PNP0A08:00/device:4a/PNP0C09:00/ACPI0003:00/power_supply/AC
lrwxrwxrwx 1 root root 0 May  4 12:55 BAT0 -> ../../devices/LNXSYSTM:00/LNXSYBUS:00/PNP0A08:00/device:4a/PNP0C09:00/PNP0C0A:00/power_supply/BAT0
```

**Comment by nkcyham** [comment]
Hi, thanks for the reply. OK about backtics, I must study protocol docs for the forum....

Yes, I can access and list the relevenat directories as ordinary user and all permissions etc are the same as yours, except that instead of AC and BAT0, I have ADP0 and BAT0 in /sys/class/power_supply/. I presume ADP0 is the adaptor I am using.

**Comment by Vaelatern** [comment]
Yet you still get `xbattmon: fopen /sys/class/power_supply/BAT0/capacity: No such file or directory` ?

**Comment by nkcyham** [comment]
Yes, I do. I will try to right a little C code to open the file, to see what happens. I know it makes little sense - unless I missed something with the permissions. I will comment again once I have more definite data.

**Comment by Vaelatern** [comment]
If you can `cat` the file as a regular user, then the file definitely exists and is readable.

**Comment by nkcyham** [comment]
@Vaelatern 

Yes, I can cat it and open it in vi and nano as a user. Must be xbattmon itself?

**Comment by Vaelatern** [comment]
`xbattmon` is cool, thanks for putting me on to that.
Works for me. Have you tried restarting, just to ensure your system is in a consistent state?

**Comment by nkcyham** [comment]
@Vaelatern 
The installation was actually done a few weeks ago but because my dog (RIP) was very ill, I left it unfinished, regarding installing some programs and configuring the system. Its just over the last few days that this was completed, but configuring tint2 failed to bring up the battery monitor. This is why I tried acpi and xbattmon, to see if they can work (acpi seems to do). The laptop has been booted several times since so it should be consistent.

Anyhow, its rather baffling but as usual the explanation is likely to be simple, if only I can find one....

**Comment by nkcyham** [comment]
@Vaelatern 
Got it!! Its a bug in the xbattmon source code. Open
`http://git.2f30.org/xbattmon/file/xbattmon.c.html`
and look at line 303. The error message string is incorrect: `PATH_BAT_CAP`. It should be of course `PATH_AC_ONLINE`. No doubt a copy-paste error, I do a lot of these myself! :-(

The problem is that `PATH_AC_ONLINE` is `#define'd as /sys/class/power_supply/AC/online` but in my system that is `/sys/class/power_supply/ADP0/online`, as I mentioned already. So I was getting `No such file or directory` for the wrong missing file. I fixed it by modifying `PATH_AC_ONLINE` #define to `/sys/class/power_supply/ADP0/online` and xbattmon works OK. Too late though - I assigned the task to conky and a fine job it does too....

I believe this change to `/sys/class/power_supply/ADP0/online` may also be the reason tint2 no longer indicates battery status. So, someone must send a bug report to xbattmon upstream. I recon a Void "official" will have a better chance than me to get through....
